static-analysis/EnumInstantiation.php

```diff
@@ -0,0 +1,48 @@
+<?php
+
+declare(strict_types=1);
+
+namespace MyCLabs\Tests\Enum\StaticAnalysis;
+
+use MyCLabs\Enum\Enum;
+
+/**
+ * @psalm-immutable
+ * @psalm-template T of 'A'|'C'
+ * @template-extends Enum<T>
+ */
+final class InstantiatedEnum extends Enum
+{
+    const A = 'A';
+    const C = 'C';
+}
+
+/**
+ * @psalm-pure
+ * @psalm-return InstantiatedEnum<'A'>
+ */
+function canCallConstructorWithConstantValue(): InstantiatedEnum
+{
+    return new InstantiatedEnum('A');
+}
+
+/**
+ * @psalm-pure
+ * @psalm-return InstantiatedEnum<'C'>
+ */
+function canCallConstructorWithConstantReference(): InstantiatedEnum
+{
+    return new InstantiatedEnum(InstantiatedEnum::C);
+}
+
+/** @psalm-pure */
+function canCallFromWithKnownValue(): InstantiatedEnum
```

Comment:
Notice how we **CANNOT** infer `InstantiatedEnum<'C'>` from this signature: this is a known problem with `Enum::from()` (static method), as I've described in the patch overview.